### ISSUE TITLE
Cannot add line to Expense Report

### ISSUE THREAD

**Issue by pstructures** [Dolibarr/dolibarr#12463]
Bug
Unable to add a line to Expense Report

## Environment
- **Version**: 10.0.3
- **OS**: Linux varda.uksrv.co.uk 3.10.0-962.3.2.lve1.5.24.5.el7.x86_64 #1 SMP Thu Nov 22 09:42:49 EST 2018 x86_64
- **Web server**: LiteSpeed
- **PHP**: 7.2.24
- **Database**: mysqli (10.2.27-MariaDB)
- **URL(s)**: /custom/mileage/mileage_card.php

## Expected and actual behavior
Expected behaviour would be for the line to be added and the php script to continue, but the actual behaviour occurs with a php 500 error.  I have the following from the error log file:
```

2019-11-17 00:09:55 DEBUG   84.9.169.96     ExpenseReport::addline qty=300, up=0.2, fk_c_type_fees=4, vatrate=0.00, date=2019-11-12, fk_project=151, type=3, comments=Something
2019-11-17 00:09:55 DEBUG   84.9.169.96     BEGIN Transaction
2019-11-17 00:09:55 INFO    84.9.169.96      getLocalTaxesFromRate vatrate=0 local=0
2019-11-17 00:09:55 DEBUG   84.9.169.96      sql=SELECT t.localtax1, t.localtax1_type, t.localtax2, t.localtax2_type, t.accountancy_code_sell, t.accountancy_code_buy FROM db_c_tva as t, db_c_country as c WHERE t.fk_pays = c.rowid AND c.code = '' AND t.taux = 0 AND t.active = 1
2019-11-17 00:09:55 INFO    84.9.169.96      get_localtax tva=0 local=1 thirdparty_buyer id=0/country_code=GB thirdparty_seller id=0/country_code=GB thirdparty_seller localtax1_assuj=0  thirdparty_seller localtax2_assuj=0
2019-11-17 00:09:55 INFO    84.9.169.96      get_localtax tva=0 local=2 thirdparty_buyer id=0/country_code=GB thirdparty_seller id=0/country_code=GB thirdparty_seller localtax1_assuj=0  thirdparty_seller localtax2_assuj=0
2019-11-17 00:09:55 INFO    84.9.169.96      --- End access to /custom/mileage/mileage_card.php
```
I have also attached the full action code below:

```
if ($action == 'beforeadd' ) {
		//setEventMessages("test message", null, 'errors');
		//$action = 'create';
		//print '<script>alert("what is going on");</script>';
		
		//Add a line to expense report
		require_once DOL_DOCUMENT_ROOT . '/expensereport/class/expensereport.class.php';
		$er = new ExpenseReport($db);
		$er->fetch(GETPOST('fk_er'));
		//Get Customer Mileage rate
		require_once DOL_DOCUMENT_ROOT.'/societe/class/societe.class.php';
		$tp = new Societe($db);
		$tp->fetch(GETPOST('fk_soc'));
		
		
		
		$mileagerate = GETPOST('mileagerate');
						$sql = "SELECT rowid,code,label,upperrate,lowerrate,cutoff,active ";
						$sql .= "FROM ".MAIN_DB_PREFIX."MileageRates";
						$sql .= " WHERE rowid = ".$mileagerate;
						$sql .= " ORDER BY rowid";
						
						$resql = $db->query($sql);
						
						if ($resql) {
							$obj = $db->fetch_object($resql);
							
							$upperrate = $obj->upperrate;
							$lowerrate = $obj->lowerrate;
							$cutoff = $obj->cutoff;
						
							$client_mileage_rate = $tp->array_options['options_clientMileageRate'];
							if ($client_mileage_rate) 
							{
								if ( strpos( $client_mileage_rate, '.' ) ) {
								$mileage_price = $client_mileage_rate;
								} else {
									$mileage_price = ($client_mileage_rate / 100);
								}
								
							} else { // If no unit price known use the dictionary value
							
								if ( strpos( $upperrate, '.' ) ) { 
									$mileage_price = floatval($upperrate);
								} else {
									$mileage_price = (floatval($upperrate) / 100);
								}
							}
		
							$er->addline(300, 0.2, 4, "0.00", "2019-11-12", "Something", 151, 1, 3, 4);
							
							if ($exprep > 0) {
								setEventMessages('Added to Expense Report Number '.$exprep, null, 'mesgs');
								
							} else {
								setEventMessages('Date is => '.print_r($er, true), null, 'errors');
								$action = 'create';
							}
		
						}
			
			foreach ($object->fields as $key => $val)
			{
				if (in_array($key, array('rowid', 'entity', 'date_creation', 'tms', 'fk_user_creat', 'fk_user_modif', 'import_key'))) continue;	// Ignore special fields

				// Set value to insert
				if (in_array($object->fields[$key]['type'], array('text', 'html'))) {
					$value = GETPOST($key, 'none');
				} elseif ($object->fields[$key]['type']=='date') {
					$value = dol_mktime(12, 0, 0, GETPOST($key.'month'), GETPOST($key.'day'), GETPOST($key.'year'));
				} elseif ($object->fields[$key]['type']=='datetime') {
					$value = dol_mktime(GETPOST($key.'hour'), GETPOST($key.'min'), 0, GETPOST($key.'month'), GETPOST($key.'day'), GETPOST($key.'year'));
				} elseif ($object->fields[$key]['type']=='price') {
					$value = price2num(GETPOST($key));
				} else {
					$value = GETPOST($key, 'alpha');
				}
				if (preg_match('/^integer:/i', $object->fields[$key]['type']) && $value == '-1') $value='';		// This is an implicit foreign key field
				if (! empty($object->fields[$key]['foreignkey']) && $value == '-1') $value='';					// This is an explicit foreign key field

				$object->$key=$value;
				if ($val['notnull'] > 0 && $object->$key == '' && ! is_null($val['default']) && $val['default'] == '(PROV)')
				{
					$object->$key = '(PROV)';
				}
				if ($val['notnull'] > 0 && $object->$key == '' && is_null($val['default']))
				{
					$error++;
					setEventMessages($langs->trans("ErrorFieldRequired", $langs->transnoentitiesnoconv($val['label'])), null, 'errors');
				}
			}

			if (! $error)
			{
				$result=$object->create($user);
				if ($result > 0)
				{
					// Creation OK
					$urltogo=$backtopage?str_replace('__ID__', $result, $backtopage):$backurlforlist;
					header("Location: ".$urltogo);
					exit;
				}
				else
				{
					// Creation KO
					if (! empty($object->errors)) setEventMessages(null, $object->errors, 'errors');
					else  setEventMessages($object->error, null, 'errors');
					$action='create';
				}
			}
			else
			{
				$action='create';
			}
		
	}
```
## Steps to reproduce the behavior

Create an ExpenseReport object and try to add a line from a custom module card.php

## [Attached files](https://help.github.com/articles/issue-attachments) (Screenshots, screencasts, dolibarr.log, debugging informations…)
[*Files*]

**Comment by atm-maxime** [comment]
Seems that you are using a custom module named mileage. If you inactivate this module, do you have any trouble ? If no, this means the issue is with mileage and you should contact the developer of this module. And close this issue since it's not a Dolibarr one.

**Comment by pstructures** [comment]
Hi, I am the developer of the module.  I put this down as a bug as it related to the core module ExpenseReports.  I have rectified the issue by adding a line manually in the custom module code rather than calling the addline function from the ExpenseReport class.  Apologies for marking this as a bug, I was just a bit unsure of where to put the information.  Is there some kind of developer discussion board?  I did see something about a slack channel but it is invitation only.

Matt
